### ISSUE TITLE
Egress fails with momentary disconnection

### ISSUE THREAD

**Issue by frostbyte73** [livekit/egress#210]
When tracks or participants drop due to connection issues (even momentarily), egress fails. 
Drops are not noticeable on user end (lasting ~100-500ms).

### Status
Completed

### Issue
When there is only one participant and there is a momentary disconnection, the template thinks there is nobody left and ends the egress

### Solution
Room closure logic fixed in https://github.com/livekit/livekit/pull/1547

### Status
~~Completed~~ Issue persists with websocket output

### Issue
Pipeline cannot handle momentary disconnection - Appsrc returns EOS early and stalls the pipeline.

### Solution
Modify appsrc to continue on EOF error due to disconnection (#305, #313)

**Comment by janoskranczler** [comment]
I tried to build on track-egress but this issue blocks me. Could you give me an ETA for it?

**Comment by janoskranczler** [comment]
Could you recommend a workaround for this issue?
The track egress always stops for me after 40-50 seconds.

**Comment by imredobos** [comment]
I experience the same issue as janoskranczler. Do you have any idea when will you able do work on this?

**Comment by frostbyte73** [comment]
@janoskranczler @imredobos looking into this today - will update when I have a fix and/or more info

**Comment by akosratku** [comment]
Hey @frostbyte73 , any ideas on workarounds for this?

**Comment by frostbyte73** [comment]
@akosratku it should be fixed for track and track composite in the latest version. A workaround for room composite would be to use a custom template, and on participant disconnection, wait a couple seconds before calling `EgressHelper.endRecording()` to make sure there are still no participants

**Comment by janoskranczler** [comment]
@frostbyte73 you mean the v1.6.0 should fix this?

**Comment by frostbyte73** [comment]
@janoskranczler taking a look

**Comment by frostbyte73** [comment]
I was able to reproduce using your script, but cannot reproduce using the latest version. I'll release now and you can try using v1.7.1

**Comment by janoskranczler** [comment]
@frostbyte73 Thank you for the fix. Unfortunately, the error still occurs for me:
```
livekit-egress-test-livekit-1  | 2023-05-10T08:11:45.676Z       ERROR   livekit service/ioinfo.go:76    egress failed   {"egressID": "EG_jmjSbVxsbFzZ", "error": "Internal data stream error."}
livekit-egress-test-livekit-1  | github.com/livekit/livekit-server/pkg/service.(*IOInfoService).UpdateEgressInfo
livekit-egress-test-livekit-1  |        /workspace/pkg/service/ioinfo.go:76
livekit-egress-test-livekit-1  | github.com/livekit/psrpc/pkg/server.(*rpcHandlerImpl[...]).handleRequest
livekit-egress-test-livekit-1  |        /go/pkg/mod/github.com/livekit/psrpc@v0.3.1-0.20230502152150-df9dd21fba11/pkg/server/rpc.go:161
livekit-egress-test-livekit-1  | github.com/livekit/psrpc/pkg/server.(*rpcHandlerImpl[...]).run.func1.1
livekit-egress-test-livekit-1  |        /go/pkg/mod/github.com/livekit/psrpc@v0.3.1-0.20230502152150-df9dd21fba11/pkg/server/rpc.go:109
```
I updated the test and fixed the versions in the docker-compose file:
https://github.com/janoskranczler/livekit-egress-test

I can constantly reproduce the issue. Have to wait for 40-50 seconds.

**Comment by janoskranczler** [comment]
@frostbyte73 Could you check it again?

**Comment by frostbyte73** [comment]
Fixed in 1.7.2

I'll leave this open for a bit, please comment if you have issues on the latest version

**Comment by janoskranczler** [comment]
Thank you for your fix. I just tested with my test project (raised the versions: livekit:v1.4.3, egress:v1.7.2). It seems that the previous error message is gone but a new error surfaced:
livekit:
```
livekit-egress-test-livekit-1  | 2023-06-08T07:42:16.284Z       ERROR   livekit service/ioinfo.go:76    egress failed   {"egressID": "EG_4SRu33Eb2P7B", "error": "internal error"}
livekit-egress-test-livekit-1  | github.com/livekit/livekit-server/pkg/service.(*IOInfoService).UpdateEgressInfo
livekit-egress-test-livekit-1  |        /workspace/pkg/service/ioinfo.go:76
livekit-egress-test-livekit-1  | github.com/livekit/psrpc/pkg/server.(*rpcHandlerImpl[...]).handleRequest
livekit-egress-test-livekit-1  |        /go/pkg/mod/github.com/livekit/psrpc@v0.3.1-0.20230528083849-53d664c6d912/pkg/server/rpc.go:161
livekit-egress-test-livekit-1  | github.com/livekit/psrpc/pkg/server.(*rpcHandlerImpl[...]).run.func1.1
livekit-egress-test-livekit-1  |        /go/pkg/mod/github.com/livekit/psrpc@v0.3.1-0.20230528083849-53d664c6d912/pkg/server/rpc.go:109
livekit-egress-test-livekit-1  | 2023-06-08T07:42:52.094Z       INFO    livekit rtc/transport.go:624    peer connection failed  {"room": "test_room", "roomID": "RM_NXG9WrnAZYMD", "participant": "EG_4SRu33Eb2P7B", "pID": "PA_Rt7zLBq6Musw", "remote": false, "transport": "SUBSCRIBER"}
livekit-egress-test-livekit-1  | 2023-06-08T07:42:52.095Z       INFO    livekit rtc/transportmanager.go:592     ignoring prefer candidate check by ICE failure because signal connection interrupted    {"room": "test_room", "roomID": "RM_NXG9WrnAZYMD", "participant": "EG_4SRu33Eb2P7B", "pID": "PA_Rt7zLBq6Musw", "remote": false, "lastSignalSince": "1m29.902795s", "signalValid": true}
livekit-egress-test-livekit-1  | 2023-06-08T07:42:52.095Z       INFO    livekit rtc/participant_signal.go:286   closing signal connection       {"room": "test_room", "roomID": "RM_NXG9WrnAZYMD", "participant": "EG_4SRu33Eb2P7B", "pID": "PA_Rt7zLBq6Musw", "remote": false}
livekit-egress-test-livekit-1  | 2023-06-08T07:42:52.095Z       INFO    livekit rtc/transport.go:1538   ice candidates  {"room": "test_room", "roomID": "RM_NXG9WrnAZYMD", "participant": "EG_4SRu33Eb2P7B", "pID": "PA_Rt7zLBq6Musw", "remote": false, "transport": "SUBSCRIBER", "lc": ["udp4 host 172.28.0.4:7882", "tcp4 host 172.28.0.4:7881", "udp4 srflx 185.199.28.180:51877 related 0.0.0.0:48847", "udp4 srflx 185.199.28.180:55258 related 0.0.0.0:34136"], "rc": ["candidate:3437248329 1 udp 2130706431 172.28.0.5 60766 typ host", "candidate:3596693351 1 udp 1694498815 185.199.28.180 57790 typ srflx raddr 0.0.0.0 rport 58191", "candidate:3596693351 1 udp 1694498815 185.199.28.180 53786 typ srflx raddr 0.0.0.0 rport 40081"], "lc (filtered)": [], "rc (filtered)": []}
livekit-egress-test-livekit-1  | 2023-06-08T07:42:52.126Z       ERROR   livekit sfu/downtrack.go:620    write rtp packet failed {"room": "test_room", "roomID": "RM_NXG9WrnAZYMD", "participant": "EG_4SRu33Eb2P7B", "pID": "PA_Rt7zLBq6Musw", "remote": false, "trackID": "TR_AM8QqZ3SqSYKah", "relayed": false, "count": 1, "error": "io: read/write on closed pipe"}
livekit-egress-test-livekit-1  | github.com/livekit/livekit-server/pkg/sfu.(*DownTrack).WriteRTP
livekit-egress-test-livekit-1  |        /workspace/pkg/sfu/downtrack.go:620
livekit-egress-test-livekit-1  | github.com/livekit/livekit-server/pkg/sfu.(*RedPrimaryReceiver).ForwardRTP.func1
livekit-egress-test-livekit-1  |        /workspace/pkg/sfu/redprimaryreceiver.go:65
livekit-egress-test-livekit-1  | github.com/livekit/livekit-server/pkg/sfu.(*DownTrackSpreader).Broadcast.func1
livekit-egress-test-livekit-1  |        /workspace/pkg/sfu/downtrackspreader.go:87
livekit-egress-test-livekit-1  | github.com/livekit/protocol/utils.ParallelExec[...]
livekit-egress-test-livekit-1  |        /go/pkg/mod/github.com/livekit/protocol@v1.5.8-0.20230601212100-a186ecb11a98/utils/parallel.go:15
livekit-egress-test-livekit-1  | github.com/livekit/livekit-server/pkg/sfu.(*DownTrackSpreader).Broadcast
livekit-egress-test-livekit-1  |        /workspace/pkg/sfu/downtrackspreader.go:86
livekit-egress-test-livekit-1  | github.com/livekit/livekit-server/pkg/sfu.(*RedPrimaryReceiver).ForwardRTP
livekit-egress-test-livekit-1  |        /workspace/pkg/sfu/redprimaryreceiver.go:64
livekit-egress-test-livekit-1  | github.com/livekit/livekit-server/pkg/sfu.(*WebRTCReceiver).forwardRTP
livekit-egress-test-livekit-1  |        /workspace/pkg/sfu/receiver.go:655
livekit-egress-test-livekit-1  | 2023-06-08T07:42:54.111Z       ERROR   livekit sfu/downtrack.go:620    write rtp packet failed {"room": "test_room", "roomID": "RM_NXG9WrnAZYMD", "participant": "EG_4SRu33Eb2P7B", "pID": "PA_Rt7zLBq6Musw", "remote": false, "trackID": "TR_AM8QqZ3SqSYKah", "relayed": false, "count": 101, "error": "io: read/write on closed pipe"}
```

egress:
```
livekit-egress-test-egress-1  | 2023-06-08T07:41:26.337Z	INFO	egress	service/handler.go:207	egress updated	{"nodeID": "NE_FfHDURU5Wxfx", "handlerID": "EGH_ATjTvHsPVwXp", "clusterID": "", "egressID": "EG_4SRu33Eb2P7B", "egressID": "EG_4SRu33Eb2P7B", "request_type": "track", "output_type": "websocket", "status": "EGRESS_ACTIVE"}
livekit-egress-test-egress-1  | panic: concurrent write to websocket connection
livekit-egress-test-egress-1  |
livekit-egress-test-egress-1  | goroutine 217 [running]:
livekit-egress-test-egress-1  | github.com/gorilla/websocket.(*messageWriter).flushFrame(0x400049e360, 0x1, {0x0?, 0x0?, 0x0?})
livekit-egress-test-egress-1  | 	/root/go/pkg/mod/github.com/gorilla/websocket@v1.5.0/conn.go:617 +0x480
livekit-egress-test-egress-1  | github.com/gorilla/websocket.(*messageWriter).Close(0x0?)
livekit-egress-test-egress-1  | 	/root/go/pkg/mod/github.com/gorilla/websocket@v1.5.0/conn.go:731 +0x4c
livekit-egress-test-egress-1  | github.com/gorilla/websocket.(*Conn).beginMessage(0x40009be160, 0x40004cfbf0, 0x9)
livekit-egress-test-egress-1  | 	/root/go/pkg/mod/github.com/gorilla/websocket@v1.5.0/conn.go:480 +0x40
livekit-egress-test-egress-1  | github.com/gorilla/websocket.(*Conn).NextWriter(0x40009be160, 0x9)
livekit-egress-test-egress-1  | 	/root/go/pkg/mod/github.com/gorilla/websocket@v1.5.0/conn.go:520 +0x44
livekit-egress-test-egress-1  | github.com/gorilla/websocket.(*Conn).WriteMessage(0x2540be400?, 0x40009b8de0?, {0x400036cc04, 0x4, 0x4})
livekit-egress-test-egress-1  | 	/root/go/pkg/mod/github.com/gorilla/websocket@v1.5.0/conn.go:773 +0x114
livekit-egress-test-egress-1  | github.com/livekit/egress/pkg/pipeline/sink.(*WebsocketSink).keepAlive(0x40009853e0)
livekit-egress-test-egress-1  | 	/workspace/pkg/pipeline/sink/websocket.go:52 +0x84
livekit-egress-test-egress-1  | created by github.com/livekit/egress/pkg/pipeline/sink.newWebsocketSink
livekit-egress-test-egress-1  | 	/workspace/pkg/pipeline/sink/websocket.go:39 +0x248
livekit-egress-test-egress-1  | 2023-06-08T07:42:16.281Z	WARN	egress	service/handler.go:195	egress failed	{"nodeID": "NE_FfHDURU5Wxfx", "clusterID": "", "egressID": "EG_4SRu33Eb2P7B", "request_type": "track", "output_type": "websocket", "error": "internal error"}
livekit-egress-test-egress-1  | 2023-06-08T07:42:16.285Z	INFO	egress	service/service_deprecated.go:33	shutting down	{"nodeID": "NE_FfHDURU5Wxfx", "clusterID": ""}
livekit-egress-test-egress-1 exited with code 0
livekit-egress-test-egress-1  | + export TMPDIR=/tmp/lkegress
livekit-egress-test-egress-1  | + TMPDIR=/tmp/lkegress
livekit-egress-test-egress-1  | + rm -rf /var/run/pulse /var/lib/pulse /root/.config/pulse
livekit-egress-test-egress-1  | + pulseaudio -D --verbose --exit-idle-time=-1 --system --disallow-exit
livekit-egress-test-egress-1  | W: [pulseaudio] main.c: Running in system mode, but --disallow-module-loading not set.
livekit-egress-test-egress-1  | N: [pulseaudio] main.c: Running in system mode, forcibly disabling SHM mode.
livekit-egress-test-egress-1  | I: [pulseaudio] main.c: Daemon startup successful.
livekit-egress-test-egress-1  | + '[' -z /tmp/lkegress ']'
livekit-egress-test-egress-1  | + mkdir -p /tmp/lkegress
livekit-egress-test-egress-1  | + exec egress
li
```

Do you have an idea what could cause this error?

**Comment by janoskranczler** [comment]
I tried it with 1.7.3 but unfortunately get this error:
```
livekit-egress-test-egress-1  | panic: runtime error: invalid memory address or nil pointer dereference
livekit-egress-test-egress-1  | [signal SIGSEGV: segmentation violation code=0x1 addr=0x368 pc=0x11f34d0]
livekit-egress-test-egress-1  | 
livekit-egress-test-egress-1  | goroutine 17 [running, locked to thread]:
livekit-egress-test-egress-1  | github.com/livekit/egress/pkg/pipeline/output.(*WebsocketOutput).SetSink.func2(0x400006ed78?)
livekit-egress-test-egress-1  |         /workspace/pkg/pipeline/output/websocket.go:74 +0xc0
livekit-egress-test-egress-1  | github.com/tinyzimmer/go-gst/gst/app.goSinkNewSampleCb.func1()
livekit-egress-test-egress-1  |         /root/go/pkg/mod/github.com/livekit/go-gst@v0.2.34-0.20230210170313-8fc9f59623d4/gst/app/cgo_exports.go:137 +0x3c
livekit-egress-test-egress-1  | github.com/tinyzimmer/go-glib/glib.(*Object).WithTransferOriginal(0x4000133020, 0x400006edb8)
livekit-egress-test-egress-1  |         /root/go/pkg/mod/github.com/livekit/go-glib@v0.0.0-20230223001336-834490045522/glib/gobject.go:343 +0x80
livekit-egress-test-egress-1  | github.com/tinyzimmer/go-gst/gst/app.goSinkNewSampleCb(0x400006ed01?, 0x4000002680?)
livekit-egress-test-egress-1  |         /root/go/pkg/mod/github.com/livekit/go-gst@v0.2.34-0.20230210170313-8fc9f59623d4/gst/app/cgo_exports.go:137 +0x88
livekit-egress-test-egress-1  | 2023-06-13T07:42:10.130Z        WARN    egress  service/handler.go:198  egress failed   {"nodeID": "NE_NFZNjwiwKBAM", "clusterID": "", "egressID": "EG_XtLcpSU4tUHL", "request_type": "track", "output_type": "websocket", "error": "internal error"}
livekit-egress-test-egress-1  | 2023-06-13T07:42:10.137Z        INFO    egress  service/service_deprecated.go:33        shutting down   {"nodeID": "NE_NFZNjwiwKBAM", "clusterID": ""}
livekit-egress-test-egress-1  | + export TMPDIR=/tmp/lkegress
livekit-egress-test-egress-1  | + TMPDIR=/tmp/lkegress
livekit-egress-test-egress-1  | + rm -rf /var/run/pulse /var/lib/pulse /root/.config/pulse
livekit-egress-test-egress-1  | + pulseaudio -D --verbose --exit-idle-time=-1 --system --disallow-exit
livekit-egress-test-egress-1  | W: [pulseaudio] main.c: Running in system mode, but --disallow-module-loading not set.
livekit-egress-test-egress-1  | N: [pulseaudio] main.c: Running in system mode, forcibly disabling SHM mode.
livekit-egress-test-egress-1  | I: [pulseaudio] main.c: Daemon startup successful.
livekit-egress-test-egress-1  | + '[' -z /tmp/lkegress ']'
livekit-egress-test-egress-1  | + mkdir -p /tmp/lkegress
livekit-egress-test-egress-1  | + exec egress
```

**Comment by frostbyte73** [comment]
that doesn't seem to be 1.7.3 - `pkg/pipeline/output/websocket.go:74` is a comment

**Comment by frostbyte73** [comment]
Fixed in v1.7.4, tested using the `websocketTestServer` from egress/test/track.go running for a few minutes

**Comment by janoskranczler** [comment]
Thank you for the fix. I tried it with the test app and now the Egress didn't stop and no errors in it, but in LiveKit server:
```
livekit-egress-test-livekit-1  | 2023-06-15T08:03:23.216Z       ERROR   livekit service/ioinfo.go:76    egress failed   {"egressID": "EG_UziLRWMTRRPe", "error": "write tcp 172.25.0.5:44820->172.25.0.2:8000: write: broken pipe"}
livekit-egress-test-livekit-1  | github.com/livekit/livekit-server/pkg/service.(*IOInfoService).UpdateEgressInfo
livekit-egress-test-livekit-1  |        /workspace/pkg/service/ioinfo.go:76
livekit-egress-test-livekit-1  | github.com/livekit/psrpc/pkg/server.(*rpcHandlerImpl[...]).handleRequest
livekit-egress-test-livekit-1  |        /go/pkg/mod/github.com/livekit/psrpc@v0.3.1-0.20230528083849-53d664c6d912/pkg/server/rpc.go:161
livekit-egress-test-livekit-1  | github.com/livekit/psrpc/pkg/server.(*rpcHandlerImpl[...]).run.func1.1
livekit-egress-test-livekit-1  |        /go/pkg/mod/github.com/livekit/psrpc@v0.3.1-0.20230528083849-53d664c6d912/pkg/server/rpc.go:109
```

The python websocket endpoint also fails with abnormal close (1006):
```
livekit-egress-test-ingest-1  | ERROR:    Exception in ASGI application
livekit-egress-test-ingest-1  | Traceback (most recent call last):
livekit-egress-test-ingest-1  |   File "/root/.cache/pypoetry/virtualenvs/livekit-egress-test-MATOk_fk-py3.10/lib/python3.10/site-packages/uvicorn/protocols/websockets/websockets_impl.py", line 254, in run_asgi
livekit-egress-test-ingest-1  |     result = await self.app(self.scope, self.asgi_receive, self.asgi_send)
livekit-egress-test-ingest-1  |   File "/root/.cache/pypoetry/virtualenvs/livekit-egress-test-MATOk_fk-py3.10/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 78, in __call__
livekit-egress-test-ingest-1  |     return await self.app(scope, receive, send)
livekit-egress-test-ingest-1  |   File "/root/.cache/pypoetry/virtualenvs/livekit-egress-test-MATOk_fk-py3.10/lib/python3.10/site-packages/fastapi/applications.py", line 276, in __call__
livekit-egress-test-ingest-1  |     await super().__call__(scope, receive, send)
livekit-egress-test-ingest-1  |   File "/root/.cache/pypoetry/virtualenvs/livekit-egress-test-MATOk_fk-py3.10/lib/python3.10/site-packages/starlette/applications.py", line 122, in __call__
livekit-egress-test-ingest-1  |     await self.middleware_stack(scope, receive, send)
livekit-egress-test-ingest-1  |   File "/root/.cache/pypoetry/virtualenvs/livekit-egress-test-MATOk_fk-py3.10/lib/python3.10/site-packages/starlette/middleware/errors.py", line 149, in __call__
livekit-egress-test-ingest-1  |     await self.app(scope, receive, send)
livekit-egress-test-ingest-1  |   File "/root/.cache/pypoetry/virtualenvs/livekit-egress-test-MATOk_fk-py3.10/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 79, in __call__
livekit-egress-test-ingest-1  |     raise exc
livekit-egress-test-ingest-1  |   File "/root/.cache/pypoetry/virtualenvs/livekit-egress-test-MATOk_fk-py3.10/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 68, in __call__
livekit-egress-test-ingest-1  |     await self.app(scope, receive, sender)
livekit-egress-test-ingest-1  |   File "/root/.cache/pypoetry/virtualenvs/livekit-egress-test-MATOk_fk-py3.10/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 21, in __call__
livekit-egress-test-ingest-1  |     raise e
livekit-egress-test-ingest-1  |   File "/root/.cache/pypoetry/virtualenvs/livekit-egress-test-MATOk_fk-py3.10/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__
livekit-egress-test-ingest-1  |     await self.app(scope, receive, send)
livekit-egress-test-ingest-1  |   File "/root/.cache/pypoetry/virtualenvs/livekit-egress-test-MATOk_fk-py3.10/lib/python3.10/site-packages/starlette/routing.py", line 718, in __call__
livekit-egress-test-ingest-1  |     await route.handle(scope, receive, send)
livekit-egress-test-ingest-1  |   File "/root/.cache/pypoetry/virtualenvs/livekit-egress-test-MATOk_fk-py3.10/lib/python3.10/site-packages/starlette/routing.py", line 341, in handle
livekit-egress-test-ingest-1  |     await self.app(scope, receive, send)
livekit-egress-test-ingest-1  |   File "/root/.cache/pypoetry/virtualenvs/livekit-egress-test-MATOk_fk-py3.10/lib/python3.10/site-packages/starlette/routing.py", line 82, in app
livekit-egress-test-ingest-1  |     await func(session)
livekit-egress-test-ingest-1  |   File "/root/.cache/pypoetry/virtualenvs/livekit-egress-test-MATOk_fk-py3.10/lib/python3.10/site-packages/fastapi/routing.py", line 289, in app
livekit-egress-test-ingest-1  |     await dependant.call(**values)
livekit-egress-test-ingest-1  |   File "/code/main.py", line 17, in websocket_endpoint
livekit-egress-test-ingest-1  |     await websocket.receive_bytes()
livekit-egress-test-ingest-1  |   File "/root/.cache/pypoetry/virtualenvs/livekit-egress-test-MATOk_fk-py3.10/lib/python3.10/site-packages/starlette/websockets.py", line 122, in receive_bytes
livekit-egress-test-ingest-1  |     self._raise_on_disconnect(message)
livekit-egress-test-ingest-1  |   File "/root/.cache/pypoetry/virtualenvs/livekit-egress-test-MATOk_fk-py3.10/lib/python3.10/site-packages/starlette/websockets.py", line 105, in _raise_on_disconnect
livekit-egress-test-ingest-1  |     raise WebSocketDisconnect(message["code"])
livekit-egress-test-ingest-1  | starlette.websockets.WebSocketDisconnect: 1006
livekit-egress-test-ingest-1  | INFO:     connection closed
```

**Comment by frostbyte73** [comment]
1006 means the websocket was closed (locally) without sending or receiving a close message. It could be related to send/receive timeouts with the python client, or related to your connection, but I can't reproduce this using a go websocket server

**Comment by frostbyte73** [comment]
It might help to have your websocket server send pings on a timer

**Comment by frostbyte73** [comment]
Closing as fixed in v1.7.5
